helix-core/src/doc_formatter.rs

```diff
@@ -0,0 +1,340 @@
+//! The `DocumentFormatter` forms the bridge between the raw document text
+//! and onscreen positioning. It yields the text graphemes as an iterator
+//! and transverses (part) of the document text. During that transversal it
+//! handles grapheme detection, softwrapping and annotation.
+//! It yields `FormattedGrapheme`s and their corresponding visual coortdinates.
+//!
+//! As both virtual text and softwrapping can insert additional lines into the document
+//! it is generally not possible to find the start of the previous visual line.
+//! Instead the `DocumentFormatter` starts at the last "checkpoint" (usually a linebreak)
+//! called a "block" and the caller must advance it as needed.
+
+use std::borrow::Cow;
+use std::fmt::Debug;
+use std::mem::{replace, take};
+
+#[cfg(test)]
+mod test;
+
+use unicode_segmentation::{Graphemes, UnicodeSegmentation};
+
+use crate::graphemes::Grapheme;
+use crate::syntax::Highlight;
+use crate::text_annotations::TextAnnotations;
+use crate::{Position, RopeGraphemes, RopeSlice};
+
+/// A preprossed Grapheme that is ready for rendering
+/// with attachted styling data
+#[derive(Debug, Clone)]
+pub struct FormattedGrapheme<'a> {
+    pub grapheme: Grapheme<'a>,
+    pub highlight: Option<Highlight>,
+    // the number of chars in the document required by this grapheme
+    pub doc_chars: u16,
+}
+
+impl<'a> FormattedGrapheme<'a> {
+    /// Returns whether this grapheme is virtual inline text
+    pub fn is_virtual(&self) -> bool {
+        // The highlight field is only used for inline virtual text
+        // so it's save to reuse that.
+        // We can not use doc_chars here as that is also 0 for the EOF space
+        let is_virtual = self.highlight.is_some();
```

Comment:
Sorry if this is a dumb question, or will become clearer if I read the whole PR, but what is highlighting if it's only for "virtual text"? I thought document text could be highlighted too. What exactly is the definition of virtual text?

Reply:
Ok so this looks a bit confusing. I anticipated that and hence the comment but maybe it's not clear enough.

The syntax highlighting is handled competently separately from the `DocumentFormatter` usually by the HighlightIterators in `helix_core::syntax`. While rendering (in `helix_term::document`) the syntax highlighting is then applied to the graphemes yielded by the `DocumentFormatter`). However the syntax highlighting code only deals in char ranges of the document so it can not apply to virtual text. Therefore the `DocumentFormatter` must emit a highlight scope for virtual text that overwrites the normal highlight scope at a certain position (it would look odd if virtual text was just highlighted the same as other text).

When I talk about virtual text I mean text that is shown on the screen but not actually part of the Document/File. In this case specifically I am only referring to inline virtual text as virtual text between lines is not actually placed by the `DocumentFormatter` and therefore not relevant here (it only reserves the necessary space for virtual lines and leaves it up to the view/rendering code to actually populate these empty lines).

The definition gets a bit tricky because there are also overlays (which are often also called virtual text). An overlay replaces one grapheme of the document with a different grapheme. This is for example used by the jump mode PR for labeling. This kind of virtual text is different because it doesn't add or remove text from the document and instead only shows one grapheme instead of the other. That means it mostly behaves like normal text so it can be selected/edited and it's part of the text so it is highlighted using the existing highlight system.

Inline virtual text needs special handling in the positioning code to ensure hence I use this property to detect if a particular grapheme is virtual text.

TLDR: Usually highlights are emitted elsewhere but that's not possible for inline virtual text so the `DocumentFormatter` needs to do it. I use that here to detect if a grapheme is part of the Document

Reply:
Ahhh, that makes a lot more sense now, thanks!

Reply:
The API was unclear tough, I have changed it to an enum so it's clear a grapheme is either virtual text with highlight or belongs to the document and takes up a fixed number of graphemes.